packages/next/src/build/segment-config/app/collect-root-param-keys.ts

```diff
@@ -0,0 +1,63 @@
+import { getSegmentParam } from '../../../server/app-render/get-segment-param'
+import type { LoadComponentsReturnType } from '../../../server/load-components'
+import type { AppPageModule } from '../../../server/route-modules/app-page/module'
+import type AppPageRouteModule from '../../../server/route-modules/app-page/module'
+import type { AppRouteModule } from '../../../server/route-modules/app-route/module'
+import {
+  isAppPageRouteModule,
+  isAppRouteRouteModule,
+} from '../../../server/route-modules/checks'
+import { InvariantError } from '../../../shared/lib/invariant-error'
+
+function collectAppPageRootParamKeys(
+  routeModule: AppPageRouteModule
+): readonly string[] {
+  let rootParams: string[] = []
+
+  let current = routeModule.userland.loaderTree
+  while (current) {
+    const [name, parallelRoutes, modules] = current
+
+    // If this is a dynamic segment, then we collect the param.
+    const param = getSegmentParam(name)?.param
+    if (param) {
+      rootParams.push(param)
+    }
+
+    // If this has a layout module, then we've found the root layout because
+    // we return once we found the first layout.
+    if (typeof modules.layout !== 'undefined') {
+      return rootParams
+    }
+
+    // This didn't include a root layout, so we need to continue.
```

Comment:
Add documentation:

You can't have a parallel route before a root layout, so we only need to go through the `children` parallel route.